### PR TITLE
[libc++] [sort] Improve performance of std::sort

### PR DESCRIPTION
**Pull request by AMP999**
Let in a wider variety of element types that are eligible to be processed by branchless sort functions, like POD structs and pointers.

This change causes `sort_stability.pass.cpp` to prefer branchless sort for `EqualType`, but not if the comparator is an arbitrary lambda. So the test is adjusted to always pass a simple comparator. `nth_element_stability.pass.cpp` and `partial_sort_stability.pass.cpp` already use a simple comparator all the time, so they don't need to be adjusted.

We considered forbidding the branchless version for unintentionally expensive comparators, e.g.
    double a[10]; std::sort(a, a+10, std::less<BigDouble>());
    Derived *a[10]; std::sort(a, a+10, std::less<VirtualBase*>());
In both cases, it is safe, but more expensive than the original PR intended (since the comparator will be invoked more often in the branchless version). However, it's invoked only 0.1% more often, and benchmarks show that the branchless version is still (much) faster despite the extra invocations. So we leave this alone.
Sample benchmarks:
- https://godbolt.org/z/x1Waz3vbY
- https://godbolt.org/z/5Ph3hfqbc

Sanity-check that "branchless sort" is used only for trivially copyable types It relies on copying the elements, so copying must be at least possible, and ideally no more expensive than move; i.e., trivial.

### CODE REVIEW

File: libcxx/test/libcxx/algorithms/sort_stability.pass.cpp [hunk 1]
```diff
@@ -68,8 +68,7 @@ void test_same() {
   auto snapshot_custom_v = v;
   std::sort(v.begin(), v.end());
   std::sort(snapshot_v.begin(), snapshot_v.end());
-  std::sort(snapshot_custom_v.begin(), snapshot_custom_v.end(),
-            [](const EqualType&, const EqualType&) { return false; });
```
Comment (mordante) [inline libcxx/test/libcxx/algorithms/sort_stability.pass.cpp:72]: Why remove this test instead of just adding the new one?

File: libcxx/include/__algorithm/sort.h [hunk 2]
```diff
@@ -144,7 +144,7 @@ template <class _Compare, class _Iter, class _Tp = typename iterator_traits<_Ite
 using __use_branchless_sort =
     integral_constant<bool,
                       __libcpp_is_contiguous_iterator<_Iter>::value && sizeof(_Tp) <= sizeof(void*) &&
-                          is_arithmetic<_Tp>::value && __is_simple_comparator<_Compare>::value>;
+                          is_trivially_copyable<_Tp>::value && __is_simple_comparator<_Compare>::value>;
```
Comment (mordante) [inline libcxx/include/__algorithm/sort.h:147]: Are there already tests for sorting trivially_copyable types?

**Review by mordante — COMMENTED** [review]
Thanks for your patch. Can you add libc++ benchmarks so we can evaluate the results?

File: libcxx/test/libcxx/algorithms/sort_stability.pass.cpp (hunk 1 above)
Comment (AMP999) [reply libcxx/test/libcxx/algorithms/sort_stability.pass.cpp:72]: The old test case passes `[](const EqualType&, const EqualType&) { return false; }` as a comparator, which is not a "simple comparator", and `__use_branchless_sort` is written in a way that is true only when the `__is_simple_comparator` is true, therefor libc++ chooses to use branchless sort for `std::sort(snapshot_v.begin(), snapshot_v.end());` but not for `std::sort(snapshot_custom_v.begin(), snapshot_custom_v.end(), [](const EqualType&, const EqualType&) { return false; });` but after the change `std::sort(snapshot_custom_v.begin(), snapshot_custom_v.end(), std::less<EqualType>())` uses branchless sort. After this change, both statements will use branchless sort while the main behavior is preserved.

**Comment by philnik777** [comment]
We've not done this before, because even comparators for trivially copyable types can be arbitrarily expensive and the types can also be arbitrarily large. Consider e.g. `string_view`. It's trivially copyable, but the comparator definitely isn't simple, which may result in _very_ bad code.

File: libcxx/test/libcxx/algorithms/sort_stability.pass.cpp (hunk 1 above)
Comment (mordante) [reply libcxx/test/libcxx/algorithms/sort_stability.pass.cpp:72]: But that means we start to strongly write our tests based on our current implementation. That's not a good idea in general. Having some duplicated test coverage can help during refactoring where different paths are taken. So I really like adding a test, I'm just less thrilled with the removal.

**Comment by AMP999** [comment]
@philnik777 I've seen you've demonstrated benchmarks on your two PRs https://reviews.llvm.org/D156956 and https://reviews.llvm.org/D151274. Could you please help me to produce benchmarks for my PR?

File: libcxx/include/__algorithm/sort.h [hunk 3]
```diff
@@ -158,6 +158,7 @@ template <class _Compare, class _RandomAccessIterator>
 inline _LIBCPP_HIDE_FROM_ABI void __cond_swap(_RandomAccessIterator __x, _RandomAccessIterator __y, _Compare __c) {
   // Note: this function behaves correctly even with proxy iterators (because it relies on `value_type`).
   using value_type = typename iterator_traits<_RandomAccessIterator>::value_type;
+  static_assert(is_trivially_copyable<value_type>::value, "");
```
Comment (cjdb) [inline libcxx/include/__algorithm/sort.h:161]: Please add a helpful message about why this is a requirement.

File: libcxx/include/__algorithm/sort.h [hunk 4]
```diff
@@ -171,6 +172,7 @@ inline _LIBCPP_HIDE_FROM_ABI void
 __partially_sorted_swap(_RandomAccessIterator __x, _RandomAccessIterator __y, _RandomAccessIterator __z, _Compare __c) {
   // Note: this function behaves correctly even with proxy iterators (because it relies on `value_type`).
   using value_type = typename iterator_traits<_RandomAccessIterator>::value_type;
+  static_assert(is_trivially_copyable<value_type>::value, "");
```
Comment (cjdb) [inline libcxx/include/__algorithm/sort.h:175]: Please add a helpful message about why this is a requirement.

File: libcxx/test/libcxx/algorithms/sort_stability.pass.cpp (hunk 1 above)
Comment (cjdb) [reply libcxx/test/libcxx/algorithms/sort_stability.pass.cpp:72]: It's worth noting that this is in `test/libcxx`, not `test/std`, and so there's more lenience for tests that are coupled with the implementation. I do think that we should avoid this practice though.

It's also worth noting that this lambda doesn't meet the requirements of `Compare` because it doesn't impose a [strict weak order](https://eel.is/c++draft/algorithms#alg.sorting.general-4) on its elements, so replacing it with something that does is a good idea.

**Review by cjdb — CHANGES_REQUESTED** [review]
Thanks for working on this. It'd be great if the benchmarks you've recorded were checked into the project (using the benchmark framework) and if you could also put the results into the commit message.

**Comment by cjdb** [comment]
@AMP999 please don't force push commits, as it messes up the review. Just pushing subsequent commits is fine: we'll squash and merge into one commit at the very end.

**Comment by AMP999** [comment]
@cjdb Sure, I will continue the review process as you instructed.

**Comment by AMP999** [comment]
@philnik777 I'd appreciate it if you ELI5 how to produce a benchmark like yours in https://github.com/llvm/llvm-project/pull/76657 with "old" and "new" comparisons. It'd be a great help for my PR. Thank you in advance.

**Comment by philnik777** [comment]
> @philnik777 I'd appreciate it if you ELI5 how to produce a benchmark like yours in #76657 with "old" and "new" comparisons. It'd be a great help for my PR. Thank you in advance.

I use an incredibly advanced technique called "I have no idea how to do it properly, so I just run the benchmark twice (on the old and new code), copy-paste the old and new numbers together and call the columns 'old' and 'new'". It's a bit wordy, but I think it gets the idea across.

**Comment by philnik777** [comment]
Oh and you get the layout like this when you run google benchmark in a terminal.

File: libcxx/test/libcxx/algorithms/sort_stability.pass.cpp (hunk 1 above)
Comment (danlark1) [reply libcxx/test/libcxx/algorithms/sort_stability.pass.cpp:72]: `return false;` imposes a perfectly correct strict weak ordering comparator

Comment (danlark1) [reply libcxx/test/libcxx/algorithms/sort_stability.pass.cpp:72]: Also, I still don't understand why we need to change the comparator here. The test checks that randomization happens for any sort you call

**Comment by philnik777** [comment]
Closing, since there hasn't been any activity for quite a while. If you want to pursue this again feel free to re-open this PR.
